### ISSUE TITLE
Reader hangs when main thread throws an exception

### ISSUE THREAD

**Issue by lonvia** [osmcode/libosmium#131]
In osm2pgsql the main thread may end hanging when throwing an exception because of an invalid location while reading nodes.

For details see https://github.com/openstreetmap/osm2pgsql/pull/455

**Comment by zerebubuth** [comment]
Hopefully this smaller test case will help track down the issue.

When I compile [this test code](https://gist.github.com/zerebubuth/cb2f397899f82f8c0947#file-main-cpp) like this: `g++ -g -O2 -std=c++11 -pthread -I../libosmium/include -o main main.cpp -lexpat -lz -lbz2` and run against a file with at least 10,000 nodes (although it might need more - I was running against a Seattle extract with 2.3M nodes), then I can reproduce the deadlock. To be precise; I've reproduced it 4 times, both at the comment line and in gdb ([gdb output](https://gist.github.com/zerebubuth/cb2f397899f82f8c0947#file-gdb_output-txt)). The libosmium version used was current master: 67e15135. The gcc version is 4.8.4-2ubuntu1~14.04, boost version is 1.54.0-4ubuntu3.1.

**Comment by zerebubuth** [comment]
I should note: the important part of reproducing it seems to be making the main thread run slower than the file reader thread. There's probably some timing issues as well, given that it only seems to work when the file reader thread is stopped on [pushing to the queue](https://github.com/osmcode/libosmium/blob/master/include/osmium/io/detail/read_thread.hpp#L80) at the time that the exception is thrown.

The exact `usleep` nonsense I've got in the test case is probably just a Skinner ritual for when I got it reproducing reliably.
